### ISSUE TITLE
Stability of Shapley Values and Multicolinearity

### ISSUE THREAD

**Issue by tadaskrisciunas** [shap/shap#288]
Hi,

First of all, thanks for a great product! I've been using Shapley-value-induced black-box model explanations for a few years now, starting with the work of Strumbelj and Kononenko (http://lkm.fri.uni-lj.si/xaigor/slo/pedagosko/dr-ui/JMLR-Strumbelj-Kononenko.pdf). Your improvements in computational efficiency in the case of decision-tree-based algorithms opens up Shapley values to more practical applications.

My colleagues and I at Zopa, a UK peer-to-peer lender and challenger bank, are in the process of adding the `shap` package to our production systems. There are a few things we want to clarify before proceeding, and after discussions we decided that contacting you is the best path forward to resolve the issues.
 
1. Does **SHAP** assume independent features? To what extent does feature correlation affect it? Is the impact different for different **SHAP** computation/estimation algorithms (e.g. less so in the tailor-made algorithm for tree ensembles)?
2. How sensitive is **SHAP** to marginal changes in the input? Our team recently attended a London DS meetup discussing **SHAP** and the participants struggled to confidently address this question.
3. For **Kernel SHAP** and similar sampling approaches, are there rigorous rules to determine necessary sample sizes and confidence intervals around **SHAP** value estimates? How can one be confident in at least the correctness of feature rankings for each datapoint, if not **SHAP** values? Also, is it the case that sample sizes may depend on the specific feature values in the datapoint?
4. Looking at the **Def. 4** (pg. 6) of Strumbelj et al. (http://lkm.fri.uni-lj.si/xaigor/slo/pedagosko/dr-ui/JMLR-Strumbelj-Kononenko.pdf), each possible feature combination from the feature space enters the sum only once. However, when computing conditional expectations from the training data, two statistically indistinguishable datapoints may enter the sum twice. We think this will not affect **SHAP** value estimates that are obtained significantly, but we wanted to double-check. Are there any implications of this for **SHAP** value sizes, and would you deem them significant?
5. Would you have any reservations about using **SHAP** values for causal inference and user recommendations? I understand this is domain-specific, but from the formulae used to calculate **SHAP** values my impression has been that they are indeed considering counterfactual outcomes via conditional expectations.

I understand that this is quite a few questions crammed into one GitHub issue, even if they are related, so let me know if you'd prefer me splitting the issue out into multiple ones.

We are looking forward to your response!

Best wishes,
Tadas

**Comment by slundberg** [comment]
Good questions! Great to hear you were already using Strumbelj and Kononenko's work, and I am glad you are finding Tree SHAP useful.

Because your questions are so good it will take me a bit of time to reply. In the meantime could you clarify what you mean in (3) by marginal changes in the input? Do you mean when the test set has a distribution shift for one feature?

**Comment by tadaskrisciunas** [comment]
@slundberg - thanks! The word "marginal" in (2)* was indeed somewhat ambiguous. Clarifying, what we meant is the following. Given a datapoint 

![image](https://user-images.githubusercontent.com/917552/47295582-e3d06280-d607-11e8-874f-f1fd6a648504.png)

we want to know if there are any guarantees the SHAP value for feature ![image](https://user-images.githubusercontent.com/917552/47295626-ff3b6d80-d607-11e8-9bf0-ae4ef5cefe9d.png) will be close to the SHAP value of the feature ![image](https://user-images.githubusercontent.com/917552/47295662-1c703c00-d608-11e8-8ffb-13c4c07376c3.png) in the datapoint

![image](https://user-images.githubusercontent.com/917552/47295723-445f9f80-d608-11e8-8866-70f6d76d2c9d.png)

with small values of ![image](https://user-images.githubusercontent.com/917552/47295746-55101580-d608-11e8-8903-dfab5d19a973.png).

There's still some vagueness in what the word _small_ means, but I hope this is enough to pinpoint the direction of the question.

--------

*: The word "marginal" was not used in (3), so I assume you meant (2).

**Comment by slundberg** [comment]
Things always take longer than I think... below are my answers:

1. We are working on writing up a careful description of the tradeoffs that are made when you use assume independence among features vs. dependence. In general if you account for feature dependence then you are guaranteed to never evaluate the behavior of your model on "invalid" data. Stated another way, accounting for feature dependence means you don't evaluate you model "away from the manifold of the training data". This is desirable because if you evaluate how a model behaves in strange unrealistic situations you are not being true to the data. However it is sometimes nice to assume independence between features so you don't assign credit to features the model doesn't use, but are correlated with features it uses. For tree models, all of this is a bit nicer since Tree SHAP guarantees to only evaluate each tree on training examples, and so they are always "on manifold". It is possible to be off manifold still because we don't force consistency with a valid training example between trees, but that is something we should release soon.
2. The SHAP values for a small change in a feature will be continuous if the model's dependence on that feature is also continuous. However, if the model has step functions in it, then of course that will also effect the SHAP values.
3. The number of samples needed for convergence in Kernel SHAP is based on the convergence of the linear model. For regularized linear models, there is a lot of work on feature selection, but all of them (as far as I am aware) depend on the specific values of a feature taken by a datapoint. The SamplingExplainer you have already been using follows the standard convergence rules of means. Not sure if that is practically helpful though. In practice you can always rerun and observe the variance.
4. If you get two samples that are the same in the training dataset are you concerned about wasted time? If there really are lots of repeated samples then those should be sampled in a way that reflects their proportion in the dataset to get a good conditional expected value estimate. Perhaps I am missing what your concern is, since I don't see a problem here.
5. I don't have any reservations about using SHAP values to point out a feature you may want to change. But if you want to know what will happen the biggest problem is actually learning a causal model. Also, if you have managed to learn a correct causal model, and you want to know the impact of making a specific prediction I would advise just checking the output of the model when you make that change.

Hope that is helpful!

**Comment by tadaskrisciunas** [comment]
Thanks. And - I'm also sorry for the long delay in response!

Your answers are very helpful. The worries we had related to (4) are partly answered by your response to (1), so they're moot.

We'll run the simulations you suggested in (3) and feed back the results here, so that others can see them too. (It may take a while, as the business is busy, but we'll do our best.) We'll also confirm your answer to (2) with a few simulations.

The write-up on the trade-offs between assuming and not assuming feature independence you mention should be very useful, so we're very keen to get hold of it! We'll follow your publications, but also, if you have a chance, let us know of the publication here.

**Comment by slundberg** [comment]
@tadaskrisciunas Thanks! I'll try to remember to post a link to any writeup we get out here.

**Comment by tmitanitky** [comment]
Hi,

the consistency and accuracy is a great idea for feature importance. I really appreciate the nice implementation.

For the Question 1, I experimented an extreme case: exactly equal two features.  In the notebook "Census income classification with XGBoost.ipynb", I changed the data-loading cell as below.
```python
X,y = shap.datasets.adult()
X_display,y_display = shap.datasets.adult(display=True)

X['Relationship_copy'] = X['Relationship']
X_display['Relationship_copy'] = X_display['Relationship']

# create a train/test split
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.2, random_state=7)
d_train = xgboost.DMatrix(X_train, label=y_train)
d_test = xgboost.DMatrix(X_test, label=y_test)
```
Then, the mean SHAP value is, 
![image](https://user-images.githubusercontent.com/40882114/51800826-cc44ed80-2278-11e9-83b6-68cf88800c6a.png)

According to the difinition of the true SHAP value, the mean SHAP values of those two features are supposed to be exactly the same and half of the mean SHAP value on 'Relationship' of the original model.  As the feature_dependence property of the explainer is "tree_path_dependent" (default), I suppose this behavior is caused by XGBoost's greedy algorithm.  It may have priority on the first passed feature, when two or more features have the same benefit (just my assumption).

I also experimented copying a continuous feature with some 'marginal change'.  The mean SHAP values are not stable, and the order of two features affected the values.  As a consequence, I suppose mean SHAP values may  not be recommended for feature selection on features with multicolinearity, at least with XGBoost.  Thanks.

**Comment by slundberg** [comment]
@tmitanitky Thanks for sharing! The issue you are seeing is indeed because of XGBoost greedy selection. Once it selects a split of the first of the two duplicate features then it will never use the second one. So...when TreeExplainer runs it will never assign any credit to the second feature because it never appears in any of the XGBoost trees.

The definition of SHAP values relies on conditional expectations, so if we don't assume feature independence then we would share the credit between the duplicated features. Even though the model does not use the second feature! But Tree SHAP only captures the dependence between features as recorded by the tree structure (otherwise is would have to learn the full joint distribution of your dataset). So this is why it does not spread out the credit in your example. As noted in my comment above, this is being "true to the model", while staying on-manifold as we go down paths in the tree. (If you want to fully model feature correlations you can do that with LinearExplainer's `feature_dependence="correlation"` option)

If you want to use XGBoost to identify feature importance in the presence of correlated features then I would recommend using column-wise subsampling during model training. This is roughly similar to Ridge regression for linear models in that it tends to evenly spread out the model over correlated features.

**Comment by tmitanitky** [comment]
@slundberg Thanks for your response! 
I will try columns-wise subsampling for correlated features 👍

**Comment by keshavramaswamy** [comment]
Hi @slundberg : Thank you very much for your work on SHAP and also your quick responses/support here!
Extending the discussion in point 2 about changes in SHAP values for a new feature created by adding marginal changes to the original feature : 
* Using any individualized model explainable method requires that two similar(or close enough) datapoints should have almost the same SHAP values
* I understand that the SHAP values changes depending on the model's dependence on the feature (continuous vs step) - but it will be great if there are any examples to show the stability of SHAP in this regard?
* For instance, this [paper](https://arxiv.org/pdf/1806.08049.pdf) shows that both SHAP and LIME can show inconsistent results when trying to explain two similar datapoints (not have similar explanations)
* I feel this concept of stability, in addition to local accuracy, missingness and consistency is also pretty important while evaluating a model interpretability method - any thoughts on this?

**Comment by huanfachen** [comment]
Hi @slundberg - thank you for your work on SHAP and your quick responses & support here. Extending the discussion in point 2, which is changes in SHAP values due to marginal changes in a feature: does SHAP global feature importance change because of the randomness in the model training? For example, when fitting a random forest or XGBoost model, random seeds are used.
From an user's perspective, an explanation of feature importance should be robust or stable to the randomness of model training. Here, the 'stability ' is also a bit vague - it could be defined as the importance orderings in the global feature importance.
Any thoughts on this?

**Comment by PhillipMaire** [comment]
@huanfachen My understanding is that because SHAP is "true to the model" as mentioned above, SHAP values could be different due to randomness in how to model is initiated. I think that will depend on the data though. if your features are (mostly) independent and all provide some (non-random) information for your predictions, then your model on each run will be consistent and therefore the SHAP values will be consistent. to quote from [this thread](https://github.com/slundberg/shap/issues/1120#issuecomment-608029803) 

> When using SHAP, remember that we are explaining the machine learning model (rather than the data!!!).

**Comment by jesuinovieira** [comment]
@CloseChoice, I believe this thread can be labeled as "good reads" (#3887)
